src/main/java/org/jnode/fs/ntfs/datarun/CompressedDataRun.java

```diff
@@ -220,43 +248,96 @@ private static int uncompressBlock(final OffsetByteArray compressed,
+            for (int i = 0; i < bitsInByte && pos < rightmostInUncompressed; i++) {
                 if ((ctag & 1) != 0) {
                     int j, lmask, dshift;
                     for (j = pos - 1, lmask = 0xFFF, dshift = 12;
                          j >= 0x10; j >>= 1) {
                         dshift--;
                         lmask >>= 1;
                     }
+
+                    // If the bit corresponding to a data element is set, the element is a two-byte compressed word.
                     final int tmp = compressed.getShort(cpos);
                     cpos += 2;
-                    final int boff = -1 - (tmp >> dshift);
-                    final int blen = Math.min(3 + (tmp & lmask), BLOCK_SIZE - pos);
 
-                    // Note that boff is negative.
-                    uncompressed.copyFrom(uncompressed, pos + boff, pos, blen);
-                    pos += blen;
+                    // While using the compressed buffers,
+                    // the stored displacement must be incremented by 1 and
+                    // the stored length must be incremented by 3,
+                    // to get the actual displacement and length.
+                    final int boff = (tmp >> dshift) + 1;
+                    int blen = (tmp & lmask) + 3;
+
+                    // Some of the bits in a flag byte might not be used.
+                    // To process compressed buffers,
+                    // the size of the compressed chunk that is stored in the chunk header MUST be used
+                    // to determine the position of the last valid byte in the chunk.
+                    // The size value MUST ignore flag bits that correspond to bytes outside the chunk.
+                    blen = Math.min(blen, rightmostInUncompressed - pos);
+
+                    // https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-xca/b1ba6d34-499c-4017-ab0c-fe2daee93efc
+                    // Lempel-Ziv compression does not require that the entirety of the data to which
+                    // a compressed word refers actually be in the uncompressed buffer when the word is processed.
+                    // In other words, it is not required that (U – displacement + length < U).
+                    // Therefore, when processing a compressed word, data MUST be copied from the start of the uncompressed target region to the end—that is,
+                    // the byte at (U – displacement) MUST be copied first, then (U – displacement + 1), and so on,
+                    // because the compressed word might refer to data that will be written during decompression.
+
+                    // Also, the Linux code is https://github.com/torvalds/linux/blob/master/fs/ntfs3/lznt.c#L275-L277 as a reference.
+
+                    // DO NOT use arrayCopy in any case.
+                    for (int k = 0; k < blen; k++) {
+                        uncompressed.put(pos, uncompressed.get(pos - boff));
```

Comment:
@tmyroadctfig 
Please ignore the other PR (which is for investigation only) 
This line is the main change. 
The uncompressed.copyFrom somehow has some problem. and it's better (and easier to be implemented) to just always copy the bytes one by one.